### DISCUSSION TITLE
Entities with array columns that work under postgres & sqlite (better-sqlite)

### DISCUSSION THREAD

**Discussion by ludwick** [mikro-orm/mikro-orm#5246] [Q&A]
Hi!

We have entities that have columns that are array types (a mixture of arrays with either numbers, booleans or strings). Mikro-orm seems to handle this perfectly fine when running against postgres, inferring the right database column type and mapping to/from Javascript. However, we currently run our unit tests (using [vitest](https://vitest.dev)) with sqlite. Under sqlite by default (and per documentation) mikro-orm creates the table schemas as text columns and when inserting records, the columns are written as comma separated strings. This works totally fine for our what we need to be testing (e.g. don't care if an array of string splits on comma). However, by default when it is loading a row into an entity, it returns the values as arrays of strings even if the declared type is an array of something not a string, e.g. `number[]`. So, for example, if we write an array `[1, 2, 3]` to a column, it writes the string `'1,2,3'`. When that column is read back it turns into `['1','2','3']` in javascript.

I think this is correct per the documentation, specifically the custom types section on [overriding the array type](https://mikro-orm.io/docs/next/custom-types#arraytype). However, we don't _need_ to override the type for postgres so I didn't want to reference the custom type in the entity class itself. I ended up doing some hacky "detect if running under vitest" stuff to get our unit tests to work correctly (that is get it to read back `[1, 2, 3]` from a loaded entity). I couldn't figure out how to really make a custom type only apply for certain database drivers or how to customize this kind of mapping just for better-sqlite (by subclassing something in that package?)

Is there a better way to do this? Obviously we could just run postgres for unit test but that changes how the CI system works and is quite a bit slower.

**Comment by B4nan** [comment]
The docs explicitly mention how to create numeric arrays:

```ts
@Property({ type: new ArrayType(i => +i), nullable: true })
numericArray?: number[];
```

https://mikro-orm.io/docs/custom-types#arraytype

Unless you use ts-morph this needs to be explicit, with ts-morph we infer the array type and do it automatically.

Or are you doing it this way already? Hard to guess, provide some code and then we can talk about why it works the way it works.

**Reply by ludwick** [reply]
I will double check the various ways we tried to do it but I thought we tried the various ways of setting property options. The issue was that for unmarshalling, the split string was ending up as a list of strings rather than numbers (but only for sqlite -- works fine for postgres, presumably because the code that is mapping it to/from the database type is different).

**Reply by ludwick** [reply]
I want to apologize for not following up on this. We constructed a way to do what we needed. Essentially at entity typescript file top-level loaded a function that inferred which driver would be in use to declare on the entity property. Some extracts from our code:

```
// File: driver.ts
function isSqliteContext() {
  return process.env['VITEST'] && process.env['TEST_DRIVER'] === 'sqlite';
}

function getSqliteArrayColumnProps() {
  return {
    boolValue: { type: BooleanArrayType },
    intValue: {
      type: new ArrayType((i) => +i),
    },
    floatValue: {
      type: new ArrayType((i) => +i),
    },
  };
}

function getPostgresArrayColumnProps() {
  return {
    boolValue: { columnType: 'boolean[]' },
    // TODO: why, when running against postgres, does this need to specify an ArrayType
    // mapper but the floatValue column with columnType of 'float8[]' does not?
    intValue: {
      type: new ArrayType((i) => +i),
      columnType: 'int8[]',
    },
    floatValue: { columnType: 'float8[]' },
  };
}

export function getArrayColumnProps() {
  if (isSqliteContext()) {
    return getSqliteArrayColumnProps();
  }
  return getPostgresArrayColumnProps();
}
```

Then in the file that declares an entity that needs these array columns, it would look something like this:

```
import { getArrayColumnProps } from './drivers';
...
const FIELD_PROPS = getArrayColumnProps();

@MikroEntity({ tableName: 'feature_values' })
export class SomeEntity extends BaseEntity {
  ...

  @Property({ nullable: true })
  stringValue?: string[];

  @Property({ ...FIELD_PROPS['boolValue'], nullable: true })
  boolValue?: boolean[];

  @Property({ ...FIELD_PROPS['intValue'], nullable: true })
  intValue?: bigint[];

  @Property({ ...FIELD_PROPS['floatValue'], nullable: true })
  floatValue?: number[];

  ...
}
```

I've been meaning to extract a generic small repro that tests this in isolation from the rest of our codebase, but haven't had teh time and this workaround worked well enough. Unfortunately I'm leaving the company where I'm using mikro-orm and may not finish that smaller repro, so wanted to post this for others to see.
